### PR TITLE
Implement MemQ S3Express Support 

### PR DESCRIPTION
**Pull request by yisheng-zhou**
## Overview  
  
This PR introduces support for MemQ S3Express, building on the existing code base created by [Ambud](https://github.com/ambud).  
  
### Key Components  
  
- **SessionCreds**: A new class responsible for holding AWS session credentials.  
  
- **SessionTokenManager**: Manages S3Express session credentials. Each bucket maintains a deque of credentials, ensuring that new credentials are added periodically while the old ones remain valid.  
  
- **S3ExpressAsyncStorageHandler**: A new handler designed to simulate the regular S3 handler, with additional session credential logic to handle file operations with the S3Express bucket. There is potential for improvement by reducing code duplication.  
  
- **S3ExpressHelper**: A utility for string processing.  
  
### Additional Changes  
  
- Minor modifications have been made to address dependency updates.  
- Adjustments have been made to replace hardcoded configurations, such as AWS region, with dynamic values.  
  
## Deployment & Testing  
  
The changes have been successfully deployed and tested within a branch in Pinterest.

### CODE REVIEW

File: memq-client/src/main/java/com/pinterest/memq/commons/storage/s3/AbstractS3StorageHandler.java [hunk 1]
```diff
@@ -73,6 +75,7 @@ public void initReader(Properties properties, MetricRegistry registry) throws Ex
       this.httpClient = new ReactorNettyRequestClient(registry);
     }
     httpClient.initialize(properties);
+    this.region = Region.of(properties.getProperty("region", "us-east-1").toLowerCase());
```
[the remark this answers is not in the document]
Comment (yisheng-zhou) [reply memq-client/src/main/java/com/pinterest/memq/commons/storage/s3/AbstractS3StorageHandler.java:78]: Make "region" to be constant. 

For the default region, I think it's OK to just put a value here. Having some metadata call is overkill. It already allows configurable region values

File: memq-client/src/main/java/com/pinterest/memq/commons/storage/s3/reader/client/RequestClient.java [hunk 2]
```diff
@@ -26,6 +27,9 @@
 
 public interface RequestClient extends Closeable {
   void initialize(Properties properties);
+  InputStream tryObjectGet(SdkHttpFullRequest request) throws IOException;
+  ByteBuf tryObjectGetAsBuffer(SdkHttpFullRequest request) throws IOException;
```
[the remark this answers is not in the document]
Comment (yisheng-zhou) [reply memq-client/src/main/java/com/pinterest/memq/commons/storage/s3/reader/client/RequestClient.java:31]: Hi Vahid, you can check line 33-34. The names are already defined. I'm making tryObjectGet and tryObjectGetAsBuffer be able to take SdkHttpFullRequest in addition.

File: memq-client/src/main/java/com/pinterest/memq/commons/storage/s3express/S3ExpressHelper.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        if (!awsRegionMap.containsKey(regionCode)) {
+            throw new S3ExpressParsingException(
+                    String.format("Unknown region code %s from bucket name %s", regionCode, bucketName));
+        }
+        return awsRegionMap.get(regionCode);
+    }
+
+    /**
+     * Get the current date in the format of "yyMMdd-HH"
+     * This is used to generate the hourly folder name. It helps the cleaning job to clean up the old data.
```
[the remark this answers is not in the document]
Comment (yisheng-zhou) [reply memq-client/src/main/java/com/pinterest/memq/commons/storage/s3express/S3ExpressHelper.java:66]: OK. I will add more notes here.

File: memq-client/src/main/java/com/pinterest/memq/commons/storage/s3express/S3ExpressAsyncStorageHandler.java [hunk 4]
```diff
@@ -0,0 +1,615 @@
...
+        try {
+          notificationSink.notify(payload, 0);
+        } catch (Exception e) {
+          notificationFailureCounter.inc();
+          throw e;
+        } finally {
+          publishTime.stop();
+        }
+      }
+      long latency = timer.stop() / NANOSECONDS_TO_SECONDS;
```
[the remark this answers is not in the document]
Comment (yisheng-zhou) [reply memq-client/src/main/java/com/pinterest/memq/commons/storage/s3express/S3ExpressAsyncStorageHandler.java:337]: That's copied from the regular handler. Will change this one.

File: memq-client/src/main/java/com/pinterest/memq/commons/storage/s3express/S3ExpressAsyncStorageHandler.java [hunk 5]
```diff
@@ -0,0 +1,615 @@
...
+    this.topic = topic;
+    initializeWriterRegistry(registry);
+    loadOutputHandlerConfigs(outputHandlerConfig, topic);
+    this.logger = Logger.getLogger(S3ExpressAsyncStorageHandler.class.getName() + "-" + topic);
+
+    this.requestExecutor = Executors.newCachedThreadPool(new DaemonThreadFactory());
+    this.executionTimer = Executors.newSingleThreadScheduledExecutor(new DaemonThreadFactory());
+    ConnectionProvider provider = ConnectionProvider.builder("s3express").maxConnections(10)
+        .maxIdleTime(Duration.ofSeconds(20)).maxLifeTime(Duration.ofSeconds(60))
+        .pendingAcquireTimeout(Duration.ofSeconds(60)).evictInBackground(Duration.ofSeconds(120))
```
[the remark this answers is not in the document]
Comment (yisheng-zhou) [reply memq-client/src/main/java/com/pinterest/memq/commons/storage/s3express/S3ExpressAsyncStorageHandler.java]: I used the values from Ambud's change

File: memq-client/src/main/java/com/pinterest/memq/commons/storage/s3express/S3ExpressHelper.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+     * Until 2024/11/05, AWS S3Express does not support object lifecycle policy. We need to clean up the old data manually.
+     * With this setup, we can easily clean up the old data by deleting the hourly folder.
```
Comment (vahidhashemian) [inline memq-client/src/main/java/com/pinterest/memq/commons/storage/s3express/S3ExpressHelper.java:68]: For S3 related comments, etc, please replace the usage of `folder` with `prefix`.

File: memq-client/src/main/java/com/pinterest/memq/commons/storage/s3express/S3ExpressHelper.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (vahidhashemian) [reply memq-client/src/main/java/com/pinterest/memq/commons/storage/s3express/S3ExpressHelper.java:66]: Once AWS provides the cleanup functionality, do you think the same placement arrangement would work? Specially if we need to set up different retention values for different topics?

[the remark this answers is not in the document]
Comment (yisheng-zhou) [reply memq-client/src/main/java/com/pinterest/memq/commons/storage/s3express/S3ExpressHelper.java]: Implement S3ExpressObjectKeyGenerator. The new OOP setup allows engineers to override S3ExpressObjectKeyGenerator.

File: memq-client/src/main/java/com/pinterest/memq/commons/storage/s3express/S3ExpressHelper.java (hunk 6 above)
Comment (yisheng-zhou) [reply memq-client/src/main/java/com/pinterest/memq/commons/storage/s3express/S3ExpressHelper.java]: Done

**Review by vahidhashemian — APPROVED** [review]

**Pull request merged**